cpp/src/arrow/meson.build

```diff
@@ -371,25 +371,38 @@ install_headers(
 )
 
 if needs_tests
-    filesystem_dep = dependency(
+    boost_dep = dependency(
         'boost',
         modules: ['filesystem'],
```

Comment:
At least on Ubuntu it appears that installation of libboost-filesystem-dev includes the other components needed for building (asio and process). However, adding those to the `modules` here will fail.

I'm not familiar with how Boost is distributed so this may or may not be a bug with Meson

Reply:
In CMake, we can use separated CMake targets for Boost. For example, `Boost::filesystem` and `Boost::process`. Can we do this with Meson?

Some libraries (e.g. Gandiva) need only `Boost::headers` but other libraries (e.g. `arrow_testing`) need `Boost::headers` and `Boost::process`.
I don't want to link needless libraries as much as possible. (e.g. I don't want to link `Boost::process` to Gandiva.)

Reply:
Absolutely. I think what is on main is closer to that, so I'll update this to reflect that. Each CMake target should be wrapped by Meson, so you can have filesystem_dep, process_dep, etc... just like in the else branch

Reply:
To clarify the update code, with Meson a call to:

```python
boost_dep = declare_dependency('boost')
```

Will return a dependency to the header only library for boost. If you additionally call with a module like:

```python
filesystem_dep = dependency('boost', modules: 'filesystem')
```

The dependency will then cause linkage against the specified module(s)